src/macros-by-example.adoc

```diff
@@ -0,0 +1,821 @@
+{ a:1, (:values {}), z:3 }          ⇒ { a:1, z:3 }
+{ a:1, (:values {b:2}), z:3 }       ⇒ { a:1, b:2, z:3 }
+{ a:1, (:values {b:2} {z:3}), z:3 } ⇒ { a:1, b:2, z:3, z:3 }
+
+{ a:1, (:values key "value") } ⇒ _**error**: struct expected for inlining into struct_
+----
+
+
+==== Inlining in template expressions
+
+The preceding examples demonstrate inlining of E-expressions into encoded data, but similar
+stream-inlining occurs within the template language, making it trivial to convert a stream to a
+list:
+
+[{mrk}]
+----
+(*public* int_list
+  [(**int\...** vals)]
+  [ vals ])
+(*public* clumsy_bag
+  [(**any\...** elts)]
+  { '': elts })
+----
+----
+(:int_list)   ⇒ []
+(:clumsy_bag) ⇒ {}
+
+(:int_list 1 2 3)    ⇒ [1, 2, 3]
+(:clumsy_bag true 2) ⇒ {'':true, '':2}
+----
+
+As a reminder, streams and lists are different, there’s no flattening involved, and declared
+types are verified:
+
+[{mrk}]
+----
+(:int_list 1 [2] 3) ⇒ _**error**: [2] is not an int_
+----
+
+TODO: demonstrate inlining in TDL macro invocations
+
+
+=== Mapping Templates Over Streams
+
+Another way to produce a stream is via a mapping form. The `*for_each*` special form evaluates a
+template once for each value provided by a macro parameter.  Each time, the parameter name is
+shadowed and rebound to the next value on the stream.
+
+[{mrk}]
+----
+(*public* prices
+  [(*symbol* currency), (*number\...* amount)]
+  (*for_each* amount
+    (price amount currency)))
+----
+----
+(:prices GBP 10 9.99 12.)
+⇒ {amount:10, currency:GBP} {amount:9.99, currency:GBP} {amount:12., currency:GBP}
+----
+
+TODO: Consider the more-general binding form
+
+
+=== Empty Streams: `*void*`
+
+To wrap up our discussion of streams, we have another important built-in macro: `void` accepts
+no values and produces no value. In other words, it’s a constant, the empty stream:
+
+----
+(:int_list (:void)) ⇒ []
+(:int_list 1 (:void) 2) ⇒ [1, 2]
+[(:void)]   ⇒ []
+{a:(:void)} ⇒ {}
+----
+
+When used as a macro argument, a `void` invocation (like any other expression) counts as one
+argument, albeit one with no value(s).
+
+[{mrk}]
+----
+(:pi (:void)) ⇒ _**error**: 'pi' expects 0 arguments, given 1_
+----
+
+Void is important because we need something outside the Ion data model to express that Ion data
+doesn't exist.  We can’t use null because `null` is a value!  In much the same way that Ion
+differentiates a null-field from an absent field (for example, `{a:null}` versus `{}`), macros
+need to differentiate the absence of value from all other values in the Ion data model.
+
+We’ll use the term _void_ most accurately as a synonym for “empty stream”, but you might catch us
+talking as if void were a value, when it’s actually no value at all.  We’ll even mint the word
+_voidable_ to describe parameters that can accept zero values, like the ``*\...*``s above. This
+neologism will prove useful, you’ll see.
+
+The special-case E-expression `(:)` is synonymous with `(:void)`
```

Comment:
Do you think it's necessary to have three synonyms—`(:values)`, `(:void)`, and `(:)`?

Reply:
They aren't really synonymous, since `void` accepts no arguments.   At one point we got rid of `void` but we brought it back once we realized how useful it was to have the dedicated term, and to have the code match that term.

Also note that `values` is over-represented in the document, I don't think it'll get used in practice anywhere near as commonly as its used herein.  You rarely need to pass more-than-one value in a context where a single argument is required.  So mostly its needed to collate a result stream.

This does make me think that perhaps this doc shouldn't make so much use of `values`...